internal/ssdownload/download_manager.go

```diff
@@ -0,0 +1,243 @@
+}
+
+func NewDownloadManager(ctx context.Context, di database.Plugin, ss sharedstorage.Plugin, dx dataexchange.Plugin, om operations.Manager, cb Callbacks) (Manager, error) {
+	if di == nil || dx == nil || ss == nil || cb == nil {
+		return nil, i18n.NewError(ctx, i18n.MsgInitializationNilDepError)
+	}
+
+	dmCtx, cancelFunc := context.WithCancel(ctx)
+	dm := &downloadManager{
+		ctx:                        dmCtx,
+		cancelFunc:                 cancelFunc,
+		database:                   di,
+		sharedstorage:              ss,
+		dataexchange:               dx,
+		operations:                 om,
+		callbacks:                  cb,
+		broadcastBatchPayloadLimit: config.GetByteSize(config.BroadcastBatchPayloadLimit),
+		workerCount:                config.GetInt(config.DownloadWorkerCount),
+		retryMaxAttempts:           config.GetInt(config.DownloadRetryMaxAttempts),
+		retryInitDelay:             config.GetDuration(config.DownloadRetryInitDelay),
+		retryMaxDelay:              config.GetDuration(config.DownloadRetryMaxDelay),
+		retryFactor:                config.GetFloat64(config.DownloadRetryFactor),
+	}
+	// Work queue is twice
+	workQueueLength := config.GetInt(config.DownloadWorkerQueueLength)
+	if workQueueLength <= 0 {
+		workQueueLength = 2 * dm.workerCount
+	}
+	if dm.retryMaxAttempts <= 0 {
+		dm.retryMaxAttempts = 1
+	}
+	dm.work = make(chan *downloadWork, workQueueLength)
+
+	dm.operations.RegisterHandler(ctx, dm, []fftypes.OpType{
+		fftypes.OpTypeSharedStorageDownloadBatch,
+		fftypes.OpTypeSharedStorageDownloadBlob,
+	})
+
+	return dm, nil
+}
+
+func (dm *downloadManager) Start() error {
+	dm.workers = make([]*downloadWorker, dm.workerCount)
+	for i := 0; i < dm.workerCount; i++ {
+		dm.workers[i] = newDownloadWorker(dm, i)
+	}
+	dm.recoveryComplete = make(chan struct{})
+	go dm.recoverDownloads(fftypes.Now())
+	return nil
+}
+
+func (dm *downloadManager) Name() string {
+	return "SharedStorageDownloadManager"
+}
+
+func (dm *downloadManager) WaitStop() {
+	dm.cancelFunc()
+	for _, w := range dm.workers {
+		<-w.done
+	}
+}
+
+func (dm *downloadManager) calcDelay(attempts int) time.Duration {
+	delay := dm.retryInitDelay
+	for i := 0; i < attempts; i++ {
+		delay = time.Duration(math.Ceil(float64(delay) * dm.retryFactor))
+	}
+	if delay > dm.retryMaxDelay {
+		delay = dm.retryMaxDelay
+	}
+	return delay
+}
+
+// recoverDownloads grabs all pending operations on startup, to restart them
+func (dm *downloadManager) recoverDownloads(startupTime *fftypes.FFTime) {
+
+	defer close(dm.recoveryComplete)
+	recovered := 0
+	pageSize := uint64(25)
+	page := uint64(0)
+	errorAttempts := 0
+	for {
+		fb := database.OperationQueryFactory.NewFilter(dm.ctx)
+		filter := fb.And(
+			fb.In("type", []driver.Value{
+				fftypes.OpTypeSharedStorageDownloadBatch,
+				fftypes.OpTypeSharedStorageDownloadBlob,
+			}),
+			fb.Eq("status", fftypes.OpStatusPending),
+			fb.Lt("created", startupTime), // retry is handled completely separately
```

Comment:
Are we worried about querying based on time? If the clock moved backwards would this potentially miss some items?

Definitely an edge case, but want to at least track it further if needed.

Reply:
Don't have a significantly better solution here off the top of my head.
Think the double edge case, of clock skew, combined with a full runtime restart, feels acceptable risk for now.